File: app/src/main/java/ua/com/foxminded/locationtrackera/ui/tracker/dialog/TrackerDialogFragment.java
package ua.com.foxminded.locationtrackera.ui.tracker.dialog;

import android.app.Dialog;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;

import ua.com.foxminded.locationtrackera.mvi.HostedDialogFragment;
import ua.com.foxminded.locationtrackera.ui.tracker.TrackerFragment;
import ua.com.foxminded.locationtrackera.ui.tracker.dialog.state.TrackerDialogScreenEffect;
import ua.com.foxminded.locationtrackera.ui.tracker.dialog.state.TrackerDialogScreenState;

public class TrackerDialogFragment extends HostedDialogFragment<
        TrackerDialogContract.View,
        TrackerDialogScreenState,
        TrackerDialogScreenEffect,
        TrackerDialogContract.ViewModel,
        TrackerDialogContract.Host>
        implements TrackerDialogContract.View {

    private static final String ARG_TYPE = "arg_type";
    private static final String DIALOG_TEXT_STRING_ID = "dialog_text";
    private static final String NEGATIVE_BUTTON_TEXT_STRING_ID = "negative_button";
    private static final String POSITIVE_BUTTON_TEXT_STRING_ID = "positive_button";

    public static TrackerDialogFragment newInstance(
            int argType,
            int dialogText,
            int negativeButtonText,
            int positiveButtonText
    ) {
        final Bundle b = new Bundle();
        b.putInt(ARG_TYPE, argType);
        b.putInt(DIALOG_TEXT_STRING_ID, dialogText);
        b.putInt(NEGATIVE_BUTTON_TEXT_STRING_ID, negativeButtonText);
        b.putInt(POSITIVE_BUTTON_TEXT_STRING_ID, positiveButtonText);
        final TrackerDialogFragment f = new TrackerDialogFragment();
        f.setArguments(b);
        return f;
    }

    @Override
    protected TrackerDialogContract.ViewModel createModel() {
        return null;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final int argType = getArguments().getInt(ARG_TYPE);
        final int message = getArguments().getInt(DIALOG_TEXT_STRING_ID);
        final int negativeButton = getArguments().getInt(NEGATIVE_BUTTON_TEXT_STRING_ID);
        final int positiveButton = getArguments().getInt(POSITIVE_BUTTON_TEXT_STRING_ID);

        return new AlertDialog.Builder(requireActivity())
                .setTitle(message)
                .setNegativeButton(negativeButton, (dialogInterface, which)
                        -> ((TrackerFragment) getParentFragment()).doNegativeButton(argType))
                .setPositiveButton(positiveButton, (dialogInterface, which)
                        -> ((TrackerFragment) getParentFragment()).doPositiveButton(argType))
                .create();
    }
}

File: app/src/main/java/ua/com/foxminded/locationtrackera/ui/tracker/dialog/TrackerDialogFragment.kt
package ua.com.foxminded.locationtrackera.ui.tracker.dialog

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import ua.com.foxminded.locationtrackera.mvi.fragments.HostedDialogFragment
import ua.com.foxminded.locationtrackera.ui.tracker.TrackerFragment
import ua.com.foxminded.locationtrackera.ui.tracker.dialog.state.TrackerDialogScreenEffect
import ua.com.foxminded.locationtrackera.ui.tracker.dialog.state.TrackerDialogScreenState

class TrackerDialogFragment : HostedDialogFragment<
        TrackerDialogContract.View,
        TrackerDialogScreenState,
        TrackerDialogScreenEffect,
        TrackerDialogContract.ViewModel,
        TrackerDialogContract.Host>(),
    TrackerDialogContract.View {

    override fun createModel(): TrackerDialogContract.ViewModel? {
        return null
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val argType: Int = arguments!!.getInt(ARG_TYPE)
        val message: Int = arguments!!.getInt(DIALOG_TEXT_STRING_ID)
        val negativeButton: Int = arguments!!.getInt(NEGATIVE_BUTTON_TEXT_STRING_ID)
        val positiveButton: Int = arguments!!.getInt(POSITIVE_BUTTON_TEXT_STRING_ID)
        return AlertDialog.Builder(requireActivity())
            .setTitle(message)
            .setNegativeButton(negativeButton) { _: DialogInterface?, _: Int ->
                (parentFragment as TrackerFragment).doNegativeButton(argType)
            }
            .setPositiveButton(positiveButton) { _: DialogInterface?, _: Int ->
                (parentFragment as TrackerFragment).doPositiveButton(argType)
            }
            .create()
    }

    companion object {
        private const val ARG_TYPE = "arg_type"
        private const val DIALOG_TEXT_STRING_ID = "dialog_text"
        private const val NEGATIVE_BUTTON_TEXT_STRING_ID = "negative_button"
        private const val POSITIVE_BUTTON_TEXT_STRING_ID = "positive_button"

        fun newInstance(
            argType: Int,
            dialogText: Int,
            negativeButtonText: Int,
            positiveButtonText: Int
        ): TrackerDialogFragment {
            val b = Bundle()
            b.putInt(ARG_TYPE, argType)
            b.putInt(DIALOG_TEXT_STRING_ID, dialogText)
            b.putInt(NEGATIVE_BUTTON_TEXT_STRING_ID, negativeButtonText)
            b.putInt(POSITIVE_BUTTON_TEXT_STRING_ID, positiveButtonText)
            val f = TrackerDialogFragment()
            f.arguments = b
            return f
        }
    }
}
